### DISCUSSION TITLE
socket.write(string) and LONG STRING cutting

### DISCUSSION THREAD

**Discussion by Phazeus** [nodejs/help#4965] [Q&A]
### Version

v16.10.0

### Platform

Linux dhcppc0 5.9.0-pf7 #14 SMP Fri Mar 25 02:16:06 +04 2022 x86_64 Intel(R) Core(TM) i3 CPU U 380 @ 1.33GHz GenuineIntel GNU/Linux

### What steps will reproduce the bug?

The sent strings are often truncated. Here is an example:

```js
var s = "PRIVMSG Phazeus :@globalwarzone > Очередные грузы по линии МОО Вече доехали по назначению - батальон 1539 Народной Милиции ДНР получил военное имущество необходимое для успешного выполнения боевых задач. Отдельная благодарность от командования подразделения братьям из Македонии за предоставленную помощь! Сбор продолжается - кто может помочь реквизиты и контакты по ссылке \r\n";
this.conn.write(s);
```

and got: `@globalwarzone > Очередные грузы по линии МОО Вече доехали по назначению - батальон 1539 Народной Милиции ДНР получил военное имущество необходимое для успешного выполнения боевых задач. Отдельная благодарность от командования подразделения братьям из`?

Only part of the string was sent.

### How often does it reproduce? Is there a required condition?

always

### What is the expected behavior?

Whole string was sent.

### What do you see instead?

`@globalwarzone > Очередные грузы по линии МОО Вече доехали по назначению - батальон 1539 Народной Милиции ДНР получил военное имущество необходимое для успешного выполнения боевых задач. Отдельная благодарность от командования подразделения братьям из Македонии за предоставленную помощь! Сбор продолжается - кто может помочь реквизиты и контакты по ссылке `

### Additional information

Encoding is  `utf8`.

```js
self.conn = net.createConnection({ host: 'irc.rizon.net', port: 6667 }, self._connectionHandler.bind(self));
```

**Comment by bnoordhuis** [comment]
Converted to a discussion because that's not a node bug, just how TCP works. TCP is byte-oriented, not message-oriented.

**Reply by Phazeus** [reply]
Sure, my messenger don't cut my messages, unlike node net module...

**Reply by bnoordhuis** [reply]
And an apple is just like an orange. Except for the color, smell, taste, texture and the fact it's a separate species, of course.

It sounds like you have some catching up to do on the TCP protocol. Your favorite search engine can point you in the right direction.

**Reply by Phazeus** [reply]
I know how this protocol works. I meant that the 'net' library does not allow you to work with it in the way you need. And how to send the long data? Splitting the data doesn't works...

**Reply by bnoordhuis** [reply]
Streams are what you want to work with. There are a gazillion npm packages for making that easier.
